pymtl3/stdlib/test/stateful/test_wrapper.py

```diff
@@ -0,0 +1,258 @@
+def gen_adapter( rtl, method_spec ):
+  initialize_args = ""
+  update_args = ""
+  assign_args = ""
+
+  for arg in method_spec.args:
+    initialize_args += """
+    s.{name}_{arg} = 0
+""".format(
+        arg=arg, name=method_spec.method_name )
+
+    update_args += """
+      s.{name}_rtl.{arg} = s.{name}_{arg}
+""".format(
+        arg=arg, name=method_spec.method_name )
+
+    assign_args += """
+    s.{name}_{arg} = {arg}
+""".format(
+        arg=arg, name=method_spec.method_name )
+
+  if method_spec.rets:
+    # has ret values
+    update_rets = ""
+    initialize_rets = ""
+
+    for ret in method_spec.rets:
+      update_rets += """
+      s.{name}_{ret} = s.{name}_rtl.{ret}
+""".format(
+          ret=ret, name=method_spec.method_name )
+
+      initialize_rets += """
+    s.{name}_{ret} = 0
+""".format(
+          ret=ret, name=method_spec.method_name )
+
+    rets = ",".join([
+        "s.{name}_{ret}".format( name=method_spec.method_name, ret=ret )
+        for ret in method_spec.rets.keys()
+    ] )
+
+    # template for generated adapter
+    # three upblk: args + call, rdy, ret
+    tmpl = """
+class RTL2CL( Component ):
+
+  def construct( s, {name} ):
+
+    {name}_rtl = deepcopy( {name} )
+    {name}_rtl._dsl.constructed = False
+    s.{name}_rtl = {name}_rtl.inverse()
+
+    s.{name}_called = False
+    s.{name}_rdy = False
+{initialize_args}
+{initialize_rets}
+
+    @s.update
+    def update_{name}_adapter():
+      s.{name}_rtl.en = Bits1( 1 ) if s.{name}_called else Bits1( 0 )
+      s.{name}_called = False
+{update_args}
+
+    @s.update
+    def update_{name}_adapter_rdy():
+      s.{name}_rdy = True if s.{name}_rtl.rdy else False
+
+    @s.update
+    def update_{name}_adapter_ret():
+{update_rets}
+
+    s.add_constraints(
+        U( update_{name}_adapter_ret ) < M( s.{name} ),
+        U( update_{name}_adapter_rdy ) < M( s.{name} ),
+        U( update_{name}_adapter ) < RD( s.{name}_rtl.en ),
+        M( s.{name} ) < U( update_{name}_adapter ) )
+
+  @non_blocking( lambda s: s.{name}_rdy )
+  def {name}( s, {args} ):
+    s.{name}_called = True
+{assign_args}
+    return {rets}
+""".format(
+        name=method_spec.method_name,
+        args=",".join( method_spec.args.keys() ),
+        initialize_args=initialize_args,
+        assign_args=assign_args,
+        update_args=update_args,
+        rets=rets,
+        update_rets=update_rets,
+        initialize_rets=initialize_rets )
+
+  else:
+    # no ret values
+    tmpl = """
+class RTL2CL( Component ):
+
+  def construct( s, {name} ):
+
+    {name}_rtl = deepcopy( {name} )
+    {name}_rtl._dsl.constructed = False
+    s.{name}_rtl = {name}_rtl.inverse()
+
+    s.{name}_called = False
+    s.{name}_rdy = False
+{initialize_args}
+
+    @s.update
+    def update_{name}_adapter():
+      s.{name}_rtl.en = Bits1( 1 ) if s.{name}_called else Bits1( 0 )
+      s.{name}_called = False
+{update_args}
```

Comment:
Add some comments saying we have to generate the adapter because setattr doesn't work for implicit value constraints. We have to compile source code to supply precise names of the arguments statically.